### ISSUE TITLE
Diagnostics for invalid `await` expressions

### ISSUE THREAD

**Issue by sharkdp** [astral-sh/ty#919]
We currently don't emit any errors in `await xyz` expressions if `xyz` is not awaitable. We should probably do something similar to what we do for iterators or context managers, where we emit helpful diagnostics if calls to respective dunder methods fail. For example:

```py
class NotAwaitable: ...

class NotAwaitable2:
    # not async!
    def __await__(self):
        pass

async def main():
    x = await 1  # should be an error
    y = await NotAwaitable()  # should be an error
    z = await NotAwaitable2()  # should be an error
```

The entrypoint to implement this is here: https://github.com/astral-sh/ruff/blob/27b03a9d7bf779049585bb6cfbf229f18352572d/crates/ty_python_semantic/src/types.rs#L4868

**Comment by theammir** [comment]
I'd love to tackle this one and see where it gets me.

**Comment by theammir** [comment]
What lint do we actually report for awaiting an improperly `Awaitable` type?
For context managers, everything technically falls under `invalid-context-manager` (possibly unbound or unimplemented `__(a)enter__`/`__(a)exit__`, or implemented incorrectly).

As I see it, there are 3 erroneous cases for `__await__` as well: missing, possibly unbound, implemented incorrectly (cannot be called with no extra arguments, returns a type that is known not to be a generator).
I couldn't find a dedicated lint for `await`-related errors, so does it make sense to use a combination of `unresolved-attribute`, `possibly-unbound-implicit-call` etc., or do we need a new lint?

**Comment by sharkdp** [comment]
Adding a new lint sounds like a good idea to me.
